token-metadata/program/src/processor/verification/collection_verification.rs

```diff
@@ -0,0 +1,218 @@
+        precedence: &[AuthorityType::Metadata, AuthorityType::MetadataDelegate],
+        ..Default::default()
+    })?;
+
+    // Validate that authority type is expected.
+    match authority_response.authority_type {
+        AuthorityType::Metadata | AuthorityType::MetadataDelegate => (),
+        _ => return Err(MetadataError::UpdateAuthorityIncorrect.into()),
+    }
+
+    // Destructure the collection field from the item metadata.
+    match metadata.collection.as_mut() {
+        Some(collection) => {
+            // Set item metadata collection to verified.
+            collection.verified = true;
+
+            // In the case of a sized collection, update the size on the collection parent.
+            if collection_metadata.collection_details.is_some() {
+                increment_collection_size(&mut collection_metadata, collection_metadata_info)?;
+            }
+        }
+        None => return Err(MetadataError::CollectionNotFound.into()),
+    };
+
+    // Reserialize metadata.
+    clean_write_metadata(&mut metadata, ctx.accounts.metadata_info)?;
+    Ok(())
+}
+
+#[allow(unreachable_code)]
+#[allow(unused_variables)]
+pub(crate) fn unverify_collection_v1(program_id: &Pubkey, ctx: Context<Unverify>) -> ProgramResult {
+    // Not tested so returning.
+    return Err(MetadataError::FeatureNotSupported.into());
+
+    // Assert program ownership/signers.
+
+    // Authority account is the collection authority and must be a signer.
+    assert_signer(ctx.accounts.authority_info)?;
+
+    if let Some(delegate_record_info) = ctx.accounts.delegate_record_info {
+        assert_owned_by(delegate_record_info, program_id)?;
+    }
+
+    assert_owned_by(ctx.accounts.metadata_info, program_id)?;
+
+    let collection_mint_info = ctx
+        .accounts
+        .collection_mint_info
+        .ok_or(MetadataError::MissingCollectionMint)?;
+    assert_owned_by(collection_mint_info, &spl_token::ID)?;
+
+    let collection_metadata_info = ctx
+        .accounts
+        .collection_metadata_info
+        .ok_or(MetadataError::MissingCollectionMetadata)?;
+
+    // Deserialize item metadata.
+    let mut metadata = Metadata::from_account_info(ctx.accounts.metadata_info)?;
+
+    // Destructure the collection field from the item metadata.  If there's no collection set, we
+    // can just short-circuit since there's nothing to unverify.
+    let collection = match metadata.collection.as_mut() {
+        Some(collection) => collection,
+        None => return Ok(()),
+    };
+
+    // Short-circuit if it's already unverified.
+    if !collection.verified {
+        return Ok(());
+    }
+
+    // The collection parent must be the actual parent of the collection item.
+    if collection.key != *collection_mint_info.key {
+        return Err(MetadataError::NotAMemberOfCollection.into());
+    }
+
+    // Ensure the metadata is derived from the mint.
+    assert_metadata_derivation(program_id, collection_metadata_info, collection_mint_info)?;
```

Comment:
I wrote this original code and function, and I like having clean assertion functions like this that do one thing, but thinking about it more, a much cheaper check here than doing the derivation is:

```rust
if metadata.mint != mint_info.key {
    return Err(MetadataError::MintMismatch.into());
}
```

We have already validated that the Metadata account is legitimate and belongs to Token Metadata, so we know that the mint field on it is the mint it was derived from and so we can just do a pubkey comparison instead of a derivation check.

If you like the function assertion approach this could be:

```rust
assert_metadata_matches_mint(&collection_metadata, collection_mint_info)?;
```

Where `assertions/metadata.rs` has it defined:

```rust
pub fn assert_metadata_matches_mint(metadata: &Metadata, mint_info: &AccountInfo) -> ProgramResult {
    if metadata.mint != *mint_info.key {
        return Err(MetadataError::MintMismatch.into());
    }
    Ok(())
}
```

Reply:
> We have already validated that the Metadata account is legitimate and belongs to Token Metadata
Not the collection metadata though.  We only do that in the branch when parent is not burned, i.e. data is not empty.

If the collection parent was burned, then the owner of `collection_metadata` will be System Program and its data will be empty.  So we cannot check that `collection_metadata.mint == collection_mint_info.key`.

Reply:
Ah you're right, probably best to leave that `assert_metadata_matches_mint` in place then.